localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py

```diff
@@ -57,22 +62,87 @@ def get_queue_attributes(self) -> dict:
     def event_source(self) -> str:
         return "aws:sqs"
 
+    def close(self) -> None:
+        self._shutdown_event.set()
+
+    def collect_messages(self, max_batch_size=10, max_batch_window=0, **kwargs) -> list[dict]:
+        # The number of ReceiveMessage requests we expect to be made in order to fill up the max_batch_size.
+        _total_expected_requests = (
+            max_batch_size + DEFAULT_MAX_RECEIVE_COUNT - 1
+        ) // DEFAULT_MAX_RECEIVE_COUNT
+
+        # The maximum duration a ReceiveMessage call should take, given how many requests
+        # we are going to make to fill the batch and the maximum batching window.
+        _maximum_duration_per_request = max_batch_window / _total_expected_requests
+
+        # Number of messages we want to receive per ReceiveMessage operation.
+        messages_per_receive = min(DEFAULT_MAX_RECEIVE_COUNT, max_batch_size)
+
+        def receive_message(num_messages: int = messages_per_receive):
+            start_request_t = time.monotonic()
+            response = self.source_client.receive_message(
+                QueueUrl=self.queue_url,
+                MaxNumberOfMessages=num_messages,
+                MessageAttributeNames=["All"],
+                MessageSystemAttributeNames=[MessageSystemAttributeName.All],
+            )
+            return response.get("Messages", []), time.monotonic() - start_request_t
+
+        batch = []
+        start_collection_t = time.monotonic()
+        while not self._shutdown_event.is_set():
+            # Adjust request size if we're close to max_batch_size
+            if (remaining := max_batch_size - len(batch)) < messages_per_receive:
+                messages_per_receive = remaining
+
+            # Return the messages received and the request duration in seconds.
+            try:
+                messages, request_duration = receive_message(messages_per_receive)
+            except Exception as e:
+                # If an exception is raised here, break the loop and return whatever
+                # has been collected early.
+                # TODO: Handle exceptions differently i.e QueueNotExist or ConnectionFailed should retry with backoff
+                LOG.warning(
+                    "Polling SQS queue failed: %s",
```

Comment:
nit: Can we add the queue name (`self.source_arn`) in the error message to simplify debugging